### DISCUSSION TITLE
Library for QYF-TM1638 16 button display board, that works w/Raspberry Pi Pico

### DISCUSSION THREAD

**Discussion by zen375dragon** [micropython/micropython#11768] [RP2040 / Pico]
I am unable to find a Micropython library for the QYF-TM1638 16 button display board. This version differs its other more common variant as it has 16 buttons instead of 8.

The 16 button version:
![image](https://github.com/micropython/micropython/assets/108360418/8e4493cf-6aa0-42a1-ba29-6dc857d8033b)

The 8 button version:
![image](https://github.com/micropython/micropython/assets/108360418/de30b826-0edc-49fd-9658-76bacb3e6c8a)

I had previously found libraries for this board however they were either written in Arduino or made to run via a Raspberry Pi Computer (NOT Rasperry Pi Pico).

I have the 8 button variant and its library is not compatible with this 16 button version.

Microcontroller: Raspberry Pi Pico
Board in subject:  QYF-TM1638 16 button

**Comment by mcauser** [comment]
The 8 button version: https://github.com/mcauser/micropython-tm1638

**Reply by zen375dragon** [reply]
https://github.com/mcauser/micropython-tm1638/issues/2

VojtechKlos had the same issue as mine in your repository. Have you been able to create the 16 button version yet?

**Reply by mcauser** [reply]
Now that I've found my QYF board, yes, I'll create a new driver to support it

**Comment by robhamerling** [comment]
Another version can be found here: https://gitlab.com/robhamerling/micropython-tm1638spi
This library currently supports only the board with 8 buttons. A minor change will be required to support boards with 16 buttons. The status of all buttons is in a 4 bytes buffer, by experimenting you may need to determine which bits support which buttons.

**Reply by mcauser** [reply]
Nice one. Looks like we have a lot of similar tm163x drivers 😄 I'll add yours to the Awesome MicroPython list

**Reply by zen375dragon** [reply]
Unfortunately I am not experienced enough with Micropython to adjust the driver.

@mcauser code:
https://github.com/mcauser/micropython-tm1638/blob/master/tm1638.py

```
    def _scan_keys(self):
        """Reads one of the four bytes representing which keys are pressed."""
        pressed = 0
        self.dio.init(Pin.IN, Pin.PULL_UP)
        for i in range(8):
            self.clk(0)
            if self.dio.value():
                pressed |= 1 << i
            self.clk(1)
        self.dio.init(Pin.OUT)
        return pressed
```

Could changing `for i in range(8):` to `for i in range(16):` be the first step in changing the code? I'm not sure if each button is wired the same way as the 8 button version. If they aren't, this probably won't work, and 4x4 scanning will need to be added, like in [this](https://www.youtube.com/watch?v=EpvVYyKwfjs) video.

**Reply by mcauser** [reply]
I found my QFY-TM1638 board and LED&KEY boards.

On the QFY, `_scan_keys()` should still use `range(8)`. The `keys()` method, however, is specific to the LED&KEY board button layout.

I'll add a new `keys16()` method, specifically for the QFY board, and test all of the other methods. Some are not relevant, such as the `leds()` method.

For now, if you fork the keys() method and print() the _scan_keys() values instead of adding to a single byte, it produces the following key map.
```
def keys16():
	tm.stb(0)
	tm._byte(66)
	for i in range(4):
		print(tm._scan_keys())
	tm.stb(1)
```

button | scan keys bytes
-- | --
N/A | 0,0,0,0
S1 | 4,0,0,0
S2 | 64,0,0,0
S3 | 0,4,0,0
S4 | 0,64,0,0
S5 | 0,0,4,0
S6 | 0,0,64,0
S7 | 0,0,0,4
S8 | 0,0,0,64
S9 | 2,0,0,0
S10 | 32,0,0,0
S11 | 0,2,0,0
S12 | 0,32,0,0
S13 | 0,0,2,0
S14 | 0,0,32,0
S15 | 0,0,0,2
S16 | 0,0,0,32

The layout it not as simple as the LED&KEY board. I'll massage those 4 bytes into a 16-bit value (0x0000-0xFFFF). ie.
```
S1 = 0x0001
S2 = 0x0002
S3 = 0x0004
S16 = 0x8000
```

Added new method `qyf_keys()` to my driver:
https://github.com/mcauser/micropython-tm1638/blob/master/tm1638.py#L183
It returns a 16-bit value, where each bit represents a button, SW1 = LSB

**Reply by mcauser** [reply]
It looks like the 7-segment is wired up differently too. I'll have to fork my driver to support both, or make a common ancestor.

eg1. Draw an "8." on the 7th (right most) digit:

QYF-TM1638 module
`tm.write(b'\x01\x00\x01\x00\x01\x00\x01\x00\x01\x00\x01\x00\x01\x00\x01\x00')`

LED&KEY module
`tm.write(b'\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\xff\x00')`

eg2. Draw an "8." on the 1st (left most) digit:

QYF-TM1638 module
`tm.write(b'\x80\x00\x80\x00\x80\x00\x80\x00\x80\x00\x80\x00\x80\x00\x80\x00')`

LED&KEY module
`tm.write(b'\xff\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00')`

(7-seg bits are rotated)

**Reply by mcauser** [reply]
... because the LED&KEY and QYF boards have different quad 7-segment types and are thus wired differently.
LED&KEY uses `3641AH-6 8P` which is common cathode.
QYF uses `3461BS` which is common anode.
